File: contents/stacks_and_queues/code/python/stack_queue.py
class Node:
    def __init__(self, val):
        self.value = val
        self.next = None


class Stack:
    def __init__(self):
        self.stack_list = None

    def push(self, val):
        if self.stack_list is None:
            self.stack_list = Node(val)
        else:
            temp = Node(val)
            temp.next = self.stack_list
            self.stack_list = temp

    def pop(self):
        if self.stack_list is None:
            raise ValueError("Stack is empty")
        else:
            pop_val = self.stack_list.value
            self.stack_list = self.stack_list.next
            return pop_val

    def get_top(self):
        return self.stack_list.value

    def print(self):
        print("Current stack list : ")
        temp = self.stack_list
        while temp.next:
            print(temp.value)
            temp = temp.next
        print(temp.value)


class Queue:
    def __init__(self):
        self.queue_list = None

    def enqueue(self, val):
        temp = Node(val)
        if self.queue_list is None:
            self.queue_list = temp
        else:
            curr = self.queue_list
            while curr.next:
                curr = curr.next
            curr.next = temp

    def dequeue(self):
        if self.queue_list is None:
            raise ValueError("Queue is empty")
        else:
            temp = self.queue_list
            dequeue_val = self.queue_list.value
            self.queue_list = temp.next
            return dequeue_val

    def get_front(self):
        curr = self.queue_list
        while curr.next:
            curr = curr.next
        return curr.value

    def print(self):
        temp = self.queue_list
        while temp.next:
            print(temp.value)
            temp = temp.next
        print(temp.value)


example_stack = Stack()
example_stack.push(1)
example_stack.push(2)
example_stack.push(3)
example_stack.print()
print("you just pop ", example_stack.pop())

example_stack.push(4)
example_stack.print()
print("current top => ", example_stack.get_top())

example_queue = Queue()
example_queue.enqueue(1)
example_queue.enqueue(2)
example_queue.enqueue(3)
example_queue.print()
example_queue.dequeue()
example_queue.enqueue(4)
example_queue.print()
print("current front => ", example_queue.get_front())

File: contents/stacks_and_queues/code/go/stack-queue.go
package main

import (
	"fmt"
	"os"
)

type stackNode struct {
	value interface{}
	prev  *stackNode
}

type Stack struct {
	top    *stackNode
	length int
}

// NewStack initializes a new stack.
func NewStack() *Stack {
	return &Stack{top: nil, length: 0}
}

// Push adds an element to the stack and returns the length of the stack.
func (s *Stack) Push(v interface{}) int {
	n := &stackNode{value: v, prev: s.top}
	s.top = n
	s.length++
	return s.length
}

// Pop removes and returns the last item from the stack.
func (s *Stack) Pop() interface{} {
	if s.length == 0 {
		fmt.Fprint(os.Stderr, "cannot pop items of an empty stack")
		os.Exit(1)
	}
	item := s.top
	s.top = item.prev
	s.length--
	return item.value
}

// Size returns the size of the stack.
func (s Stack) Size() int {
	return s.length
}

// Top returns the last item in the stack.
func (s Stack) Top() interface{} {
	if s.length == 0 {
		fmt.Fprint(os.Stderr, "cannot peek an empty stack")
		os.Exit(1)
	}
	return s.top.value
}

type queueNode struct {
	value interface{}
	next  *queueNode
}

type Queue struct {
	front  *queueNode
	length int
}

// NewQueue initializes a new queue.
func NewQueue() *Queue {
	return &Queue{front: nil, length: 0}
}

// Enqueue adds an item to the queue and returns the length of the queue.
func (q *Queue) Enqueue(v interface{}) int {
	n := &queueNode{value: v, next: nil}
	if q.front == nil {
		q.front = n
	} else {
		last := q.front
		for last.next != nil {
			last = last.next
		}
		last.next = n
	}
	q.length++
	return q.length
}

// Dequeue removes and returns an item from the queue.
func (q *Queue) Dequeue() interface{} {
	if q.length == 0 {
		fmt.Fprint(os.Stderr, "cannot dequeue an empty queue")
		os.Exit(1)
	}
	item := q.front
	if q.length == 1 {
		q.front = nil
	} else {
		q.front = q.front.next
	}
	q.length--
	return item.value
}

// Size returns the size of the queue.
func (q Queue) Size() int {
	return q.length
}

// Front returns the first element in the queue.
func (q Queue) Front() interface{} {
	if q.length == 0 {
		fmt.Fprint(os.Stderr, "cannot peek an empty queue")
		os.Exit(1)
	}
	return q.front.value
}

func main() {
	stack := NewStack()

	stack.Push(4)
	stack.Push(5)
	stack.Push(9)

	fmt.Println(stack.Pop())
	fmt.Println(stack.Size())
	fmt.Println(stack.Top())

	queue := NewQueue()

	queue.Enqueue(4)
	queue.Enqueue(5)
	queue.Enqueue(9)

	fmt.Println(queue.Dequeue())
	fmt.Println(queue.Size())
	fmt.Println(queue.Front())
}
